modules/ml/src/main/java/org/apache/ignite/ml/selection/cv/CrossValidation.java

```diff
@@ -311,152 +449,174 @@ public CrossValidationResult score(DatasetTrainer<M, L> trainer, Metric<L> score
 
             try (LabelPairCursor<L> cursor = testDataIterSupplier.apply(trainSetFilter, mdl)) {
-                scores[i] = scoreCalculator.score(cursor.iterator());
-            } catch (Exception e) {
+                scores[i] = metric.score(cursor.iterator());
+            }
+            catch (Exception e) {
                 throw new RuntimeException(e);
             }
         }
 
         return scores;
     }
 
-
     /**
-     * Computes cross-validated metrics with a passed parameter grid.
-     *
-     * The real cross-validation training will be called each time for each parameter set.
+     * Computes cross-validated metrics.
      *
-     * @param pipeline        Pipeline of stages.
-     * @param scoreCalculator Base score calculator.
-     * @param ignite          Ignite instance.
-     * @param upstreamCache   Ignite cache with {@code upstream} data.
-     * @param filter          Base {@code upstream} data filter.
-     * @param amountOfFolds   Amount of folds.
-     * @param paramGrid       Parameter grid.
+     * @param datasetBuilderSupplier Dataset builder supplier.
+     * @param testDataIterSupplier Test data iterator supplier.
      * @return Array of scores of the estimator for each run of the cross validation.
      */
-    public CrossValidationResult score(Pipeline<K, V, Integer, Double> pipeline,
-                                       Metric<L> scoreCalculator,
-                                       Ignite ignite,
-                                       IgniteCache<K, V> upstreamCache,
-                                       IgniteBiPredicate<K, V> filter,
-                                       int amountOfFolds,
-                                       ParamGrid paramGrid) {
+    private double[] scorePipeline(Function<IgniteBiPredicate<K, V>, DatasetBuilder<K, V>> datasetBuilderSupplier,
+        BiFunction<IgniteBiPredicate<K, V>, M, LabelPairCursor<L>> testDataIterSupplier) {
+        double[] scores = new double[amountOfFolds];
 
-        List<Double[]> paramSets = new ParameterSetGenerator(paramGrid.getParamValuesByParamIdx()).generate();
+        double foldSize = 1.0 / amountOfFolds;
+        for (int i = 0; i < amountOfFolds; i++) {
+            double from = foldSize * i;
+            double to = foldSize * (i + 1);
 
-        CrossValidationResult cvRes = new CrossValidationResult();
+            IgniteBiPredicate<K, V> trainSetFilter = (k, v) -> {
+                double pnt = mapper.map(k, v);
+                return pnt < from || pnt > to;
+            };
 
-        DatasetTrainer trainer = pipeline.getTrainer();
+            DatasetBuilder<K, V> datasetBuilder = datasetBuilderSupplier.apply(trainSetFilter);
+            PipelineMdl<K, V> mdl = pipeline.fit(datasetBuilder);
 
-        paramSets.forEach(paramSet -> {
-            Map<String, Double> paramMap = new HashMap<>();
+            try (LabelPairCursor<L> cursor = testDataIterSupplier.apply(trainSetFilter, (M)mdl)) {
```

Comment:
Do I understand correctly that we train and test a model on the same dataset (extracted using `trainSetFilter`)?